go/vt/mysqlctl/reparent.go

```diff
@@ -4,14 +4,72 @@
 
 package mysqlctl
 
+/*
+This file contains the reparenting methods for mysqlctl.
+
+TODO(alainjobart) Once refactoring is done, remove unused code paths.
+*/
+
 import (
 	"fmt"
 	"time"
 
 	log "github.com/golang/glog"
 	"github.com/youtube/vitess/go/vt/mysqlctl/proto"
+	"golang.org/x/net/context"
 )
 
+// CreateReparentJournal returns the commands to execute to create
+// the _vt.reparent_journal table. It is safe to run these commands
+// even if the table already exists.
+func CreateReparentJournal() []string {
+	return []string{
+		"CREATE DATABASE IF NOT EXISTS _vt",
+		`CREATE TABLE IF NOT EXISTS _vt.reparent_journal (
+  time_created_ns BIGINT UNSIGNED NOT NULL,
+  action_name VARCHAR(250) NOT NULL,
+  master_alias VARCHAR(32) NOT NULL,
+  replication_position VARCHAR(250) DEFAULT NULL,
+  PRIMARY KEY (time_created_ns)) ENGINE=InnoDB`}
+}
+
+// PopulateReparentJournal returns the SQL command to use to populate
+// the _vt.reparent_journal table, as well as the time_created_ns
+// value used.
+func PopulateReparentJournal(timeCreatedNS int64, actionName, masterAlias string, pos proto.ReplicationPosition) string {
+	return fmt.Sprintf("INSERT INTO _vt.reparent_journal "+
+		"(time_created_ns, action_name, master_alias, replication_position) "+
+		"VALUES (%v, '%v', '%v', '%v')",
+		timeCreatedNS, actionName, masterAlias, proto.EncodeReplicationPosition(pos))
+}
+
+// queryReparentJournal returns the SQL query to use to query the database
+// for a reparent_journal row.
+func queryReparentJournal(timeCreatedNS int64) string {
+	return fmt.Sprintf("SELECT action_name, master_alias, replication_position FROM _vt.reparent_journal WHERE time_created_ns=%v", timeCreatedNS)
+}
+
+// WaitForReparentJournal will wait until the context is done for
+// the row in the reparent_journal table.
+func (mysqld *Mysqld) WaitForReparentJournal(ctx context.Context, timeCreatedNS int64) error {
+	for {
+		qr, err := mysqld.fetchSuperQuery(queryReparentJournal(timeCreatedNS))
+		if err == nil && len(qr.Rows) == 1 {
+			// we have the row, we're done
+			return nil
+		}
+
+		// check for context being done
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
```

Comment:
If the ctx is cancelled during the Sleep below, then another fetch will be attempted before the loop quits. It should be pretty low-cost to select on a timer here instead of "default:", so we can respect the expected semantics of context cancellation (quit what you're doing and don't start anything new after Done() is closed).